linera-core/src/client.rs

```diff
@@ -746,9 +760,28 @@ where
                 .flatten()
                 .collect::<Vec<_>>();
                 if !values.is_empty() {
-                    self.process_certificate(certificate.clone(), values)
+                    self.process_certificate(certificate.clone(), values, BTreeMap::new())
                         .await?;
                 }
+            } else if let LocalNodeError::WorkerError(WorkerError::BlobsNotFound(blob_ids)) = &err {
+                let blobs = future::join_all(blob_ids.iter().map(|blob_id| {
+                    LocalNodeClient::<S>::download_blob(nodes.clone(), block.chain_id, *blob_id)
+                }))
+                .await
+                .into_iter()
+                .flatten()
+                .collect::<Vec<_>>();
+                if !blobs.is_empty() {
+                    self.process_certificate(
+                        certificate.clone(),
+                        vec![],
+                        blobs
+                            .iter()
+                            .map(|blob| (BlobId::new(blob), blob.to_owned()))
+                            .collect(),
```

Comment:
I'd prefer to treat blob hashing just like certificate value hashing. And personally I would prefer how it's currently done for the former: Have a wrapper type `HashedBlob` to memoize the hash, and make that type enforce (e.g. at deserialization, or in the `From` impl or something) that the hash is always correct. Otherwise any call site like this one is prone to errors that could make us associate the wrong ID with a blob.